httpbp/client_middlewares.go

```diff
@@ -0,0 +1,239 @@
+package httpbp
+
+import (
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/reddit/baseplate.go/breakerbp"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/sony/gobreaker"
+)
+
+// ClientMiddleware is used to build HTTP client middlewares by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(roundTripper http.RoundTripper) http.RoundTripper
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middlewares plus any additional client middlewares passed into this
+// function.
+func NewClient(config Config, middlewares ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+	defaults := []ClientMiddleware{
+		MonitorClient(config),
+		CircuitBreaker(config.CircuitBreaker),
+		Retries(config),
```

Comment:
Any thoughts on default middlewares?

Reply:
I think we should have another one that calls `ClientErrorFromResponse` before and calls `DrainAndClose(resp.Body)` after (or just defer it).